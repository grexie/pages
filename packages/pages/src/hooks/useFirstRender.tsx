import { createContext } from '@grexie/context';
import { useEffect, useState } from 'react';

export const {
  Provider: FirstRenderProvider,
  with: withFirstRenderProvider,
  use: useFirstRender,
} = createContext<boolean>('Pages.FirstRender', Provider => ({ children }) => {
  const [isFirstRender, setFirstRender] = useState(true);

  useEffect(() => {
    setFirstRender(false);
  }, []);

  return <Provider value={isFirstRender}>{children}</Provider>;
});
