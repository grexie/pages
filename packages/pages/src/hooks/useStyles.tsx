import EventEmitter from 'events';
import React, { useMemo, useState } from 'react';
import { createContextWithProps } from '../utils/context';

export interface StylesProviderProps {
  styles: StylesContext;
}

export class StylesContext extends EventEmitter {
  #styles = new Set<{ hash: string; css: string }>();

  add(hash: string, css: string) {
    const entry = { hash, css } as { hash: string; css: string };
    this.#styles.add(entry);
    this.emit('update');
    return () => {
      this.#styles.delete(entry);
      this.emit('update');
    };
  }

  [Symbol.iterator](): Iterator<{ hash: string; css: string }> {
    const stylesMap = [...this.#styles].reduce((a, { hash, css }) => {
      if (hash in a) {
        return a;
      } else {
        return { ...a, [hash]: css };
      }
    }, {}) as Record<string, string>;
    const styles = Object.keys(stylesMap).map(hash => ({
      hash,
      css: stylesMap[hash],
    }));
    return styles[Symbol.iterator]();
  }
}

export const { with: withStyles, use: _useStyles } = createContextWithProps<
  StylesContext,
  StylesProviderProps
>(Provider => ({ styles, children }) => (
  <Provider value={styles}>{children}</Provider>
));

export const useStyles = () => {
  const [, setState] = useState({});
  const styles = _useStyles();

  useMemo(() => {
    styles.on('update', () => setState({}));
  }, []);

  return styles;
};

export const Styles = () => {
  const styles = useStyles();

  return (
    <>
      {[...styles].map(({ hash, css }) => (
        <style key={hash} dangerouslySetInnerHTML={{ __html: css }} />
      ))}
    </>
  );
};
