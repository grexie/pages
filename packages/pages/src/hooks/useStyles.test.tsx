import { renderHook } from '@testing-library/react';
import { expect } from 'chai';
import { wrapStyles } from '@grexie/pages-runtime-styles';
import { StylesContext, useStyles, withStyles } from './useStyles.js';
import { withFirstRenderProvider } from './useFirstRender.js';

describe.only('useStyles', () => {
  it('should render styles in document head', () => {
    const stylesContext = new StylesContext();
    const styles = wrapStyles(
      '1',
      'css1',
      { a: '--a', b: '--b', c: '--c' },
      { '--a': '#111', '--b': '#222', '--c': '#333' }
    );

    renderHook(
      () => {
        styles.use();

        return null;
      },
      {
        wrapper: withFirstRenderProvider(
          withStyles({ styles: stylesContext })(({ children }) => {
            const styles = useStyles();
            return <>{children}</>;
          })
        ),
      }
    );
    expect(document.head).toMatchSnapshot();
  });
});
