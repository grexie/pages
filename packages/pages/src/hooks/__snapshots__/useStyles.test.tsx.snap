// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStyles should render styles in document head 1`] = `<div />`;
